src/flint/types/fmpz_mod.pyx

```diff
@@ -0,0 +1,340 @@
+from flint.flintlib.flint cimport slong
+from flint.flintlib.fmpz cimport (
+    fmpz_t,
+    fmpz_one,
+    fmpz_set,
+    COEFF_IS_MPZ,
+    fmpz_get_str,
+    fmpz_init,
+    fmpz_clear,
+    fmpz_equal
+)
+from flint.flintlib.fmpz_mod cimport *
+
+from flint.utils.typecheck cimport typecheck
+from flint.utils.conversion cimport str_from_chars
+from flint.flint_base.flint_base cimport flint_scalar
+from flint.types.fmpz cimport (
+    fmpz,
+    any_as_fmpz,
+)
+
+cimport libc.stdlib
+
+# TODO: import this from types.fmpz somehow, it's not good
+# to have the function repeated here.
+cdef fmpz_get_intlong(fmpz_t x):
+    """
+    Convert fmpz_t to a Python int or long.
+    """
+    cdef char * s
+    if COEFF_IS_MPZ(x[0]):
+        s = fmpz_get_str(NULL, 16, x)
+        v = int(str_from_chars(s), 16)
+        libc.stdlib.free(s)
+        return v
+    else:
+        return <slong>x[0]
+
+cdef class fmpz_mod_ctx:
+    """
+    """
+    def __cinit__(self, mod):
+        self.__init__(mod)
+
+    def __dealloc__(self):
+        fmpz_mod_ctx_clear(self.val)
+
+    def __init__(self, mod):
+        """
+        """
+        # Ensure modulus is fmpz type
+        if not typecheck(mod, fmpz):
+            mod = any_as_fmpz(mod)
+            if mod is NotImplemented:
+                raise NotImplementedError("TODO")
+
+        # Ensure modulus is positive
+        if mod < 1:
+            raise ValueError("Modulus is expected to be positive")
+
+        # Init the context
+        fmpz_mod_ctx_init(self.val, (<fmpz>mod).val)
+    
+    def modulus(self):
+        """
+        Return the modulus from the context as an fmpz
+        type
+        """
+        n = fmpz()
+        fmpz_set(n.val, (<fmpz_t>self.val.n))
+        return n
+
+    def __eq__(self, other):
+        if typecheck(other, fmpz_mod_ctx):
+            return fmpz_equal(self.val.n, (<fmpz_mod_ctx>other).val.n)
+        return False
+
+    def __hash__(self):
+        return hash(self.modulus())
+
+    def __repr__(self):
+        return "Context for fmpz_mod with modulus: {}".format(
+            self.modulus()
+        )
+
+    def __str__(self):
+        return "fmpz_mod_ctx({})".format(
+            self.modulus()
+        )
+
+    def __call__(self, val):
+        return fmpz_mod(val, self)
+
+# TODO: this seems stupid 
+def any_as_fmpz_mod(obj, ctx):
```

Comment:
This function is always called in a context where there is a `self` so I would make it a `cdef` method on the `fmpz_mod` class.